src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContextHelper.cs

```diff
@@ -0,0 +1,167 @@
+                        .ToImmutableArray();
+
+                    // Update the progress dialog with the count of projects to actually fix. We'll update the progress
+                    // bar as we get all the documents in AddDocumentDiagnosticsAsync.
+
+                    progressTracker.AddItems(projectsToFix.Length);
+
+                    var diagnostics = new ConcurrentDictionary<ProjectId, ImmutableArray<Diagnostic>>();
+                    using (var _ = ArrayBuilder<Task>.GetInstance(projectsToFix.Length, out var tasks))
+                    {
+                        foreach (var projectToFix in projectsToFix)
+                            tasks.Add(Task.Run(async () => await AddDocumentDiagnosticsAsync(diagnostics, projectToFix).ConfigureAwait(false), cancellationToken));
+
+                        await Task.WhenAll(tasks).ConfigureAwait(false);
+                        allDiagnostics = allDiagnostics.AddRange(diagnostics.SelectMany(i => i.Value));
+                    }
+
+                    break;
+            }
+
+            if (allDiagnostics.IsEmpty)
+            {
+                return ImmutableDictionary<Document, ImmutableArray<Diagnostic>>.Empty;
+            }
+
+            return await GetDocumentDiagnosticsToFixAsync(
+                fixAllContext.Solution, allDiagnostics, fixAllContext.CancellationToken).ConfigureAwait(false);
+
+            async Task AddDocumentDiagnosticsAsync(ConcurrentDictionary<ProjectId, ImmutableArray<Diagnostic>> diagnostics, Project projectToFix)
+            {
+                try
+                {
+                    var projectDiagnostics = await fixAllContext.GetAllDiagnosticsAsync(projectToFix).ConfigureAwait(false);
+                    diagnostics.TryAdd(projectToFix.Id, projectDiagnostics);
+                }
+                finally
+                {
+                    progressTracker.ItemCompleted();
+                }
+            }
+
+            static async Task<ImmutableDictionary<Document, ImmutableArray<Diagnostic>>> GetSpanDiagnosticsAsync(
+                FixAllContext fixAllContext,
+                IEnumerable<KeyValuePair<Document, ImmutableArray<TextSpan>>> documentsAndSpans)
+            {
+                var builder = PooledDictionary<Document, ArrayBuilder<Diagnostic>>.GetInstance();
+                foreach (var (document, spans) in documentsAndSpans)
+                {
+                    foreach (var span in spans)
+                    {
+                        var documentDiagnostics = await fixAllContext.GetDocumentSpanDiagnosticsAsync(document, span).ConfigureAwait(false);
+                        if (!builder.TryGetValue(document, out var arrayBuilder))
+                        {
+                            arrayBuilder = ArrayBuilder<Diagnostic>.GetInstance();
+                            builder.Add(document, arrayBuilder);
+                        }
+
+                        arrayBuilder.AddRange(documentDiagnostics);
```

Comment:
i believe there's already a multi-add helper for this so you don't need to do the ArrayBuilder stuff above.

Reply:
`MultiAdd` extension only works if the value being added is a single diagnostic. I have added a new extension `MultiAddRange` and used it.